File: 1.HTTP-AND-CORS/code/typescript/mime_types_the_content_type_vocabulary_1.ts
import express, { Request, Response } from 'express';

const app = express();
app.use(express.json());

interface Note {
    id: number;
    title: string;
    done: boolean;
}

app.post('/api/v1/notes', (req: Request, res: Response) => {
    // 1. read request body (express.json() handles this)
    const inNote: Note = req.body;
    
    if (!inNote || typeof inNote !== 'object') {
        res.status(400).send('invalid JSON');
        return;
    }
    
    // 2. headers FIRST
    res.setHeader('Content-Type', 'application/json');
    // 3. status
    res.status(201);
    
    inNote.id = 42;
    // 4. body LAST
    res.json(inNote);
});

app.listen(8080, () => {
    console.log('Server started on port 8080');
});

File: 1.HTTP-AND-CORS/code/rust/mime_types_the_content_type_vocabulary_1.rs
use actix_web::{web, App, HttpResponse, HttpServer, Responder};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
struct Note {
    id: Option<i32>,
    title: String,
    done: bool,
}

async fn create_note(mut in_note: web::Json<Note>) -> impl Responder {
    // 1. read request body (web::Json handles deserialization)
    
    // 2. headers FIRST (handled by HttpResponse builder)
    // 3. status
    // 4. body LAST
    in_note.id = Some(42);
    
    HttpResponse::Created()
        .content_type("application/json")
        .json(in_note.into_inner())
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| {
        App::new().route("/api/v1/notes", web::post().to(create_note))
    })
    .bind("127.0.0.1:8080")?
    .run()
    .await
}
